docs/core/extensions/logger-message-generator.md

```diff
@@ -0,0 +1,405 @@
+        Microsoft.Extensions.Logging.LogLevel logLevel,
+        Microsoft.Extensions.Logging.EventId eventId,
+        TState state,
+        System.Exception? exception,
+        Func<TState, System.Exception?, string> formatter);
+}
+```
+
+As a general rule, the first instance of `ILogger`, `LogLevel`, and `Exception` are treated specially in the log method signature of the source generator. Subsequent instances are treated like normal parameters to the message template:
+
+```csharp
+// This is a valid attribute usage
+[LoggerMessage(
+    EventId = 110, Level = LogLevel.Debug, Message = "M1 {ex3} {ex2}")]
+public static partial void ValidLogMethod(
+    ILogger logger,
+    Exception ex,
+    Exception ex2,
+    Exception ex3);
+
+// This causes a warning
+[LoggerMessage(
+    EventId = 0, Level = LogLevel.Debug, Message = "M1 {ex} {ex2}")]
+public static partial void WarningLogMethod(
+    ILogger logger,
+    Exception ex,
+    Exception ex2);
+```
+
+> [!IMPORTANT]
+> The warnings emitted provide details as to the correct usage of the `LoggerMessageAttribute`. In the preceding example the `WarningLogMethod` will report a `DiagnosticSeverity.Warning` of `SYSLIB0025`.
+>
+> ```console
+> Don't include a template for ex in the logging message since it is implicitly taken care of.
+> ```
+
+### Case-insensitive template name support
+
+The generator uses case-insensitive comparison between parameters in the message template and log message argument names so when the ILogger enumerates the state, the argument will be picked up by message template, which can make the logs nicer to consume:
+
+```csharp
+public partial class LoggingExample
+{
+    private readonly ILogger _logger;
+
+    public LoggingExample(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(
+        EventId = 10,
+        Level = LogLevel.Information,
+        Message = "Welcome to {City} {Province}!")]
+    public partial void LogMethodSupportsPascalCasingOfNames(
+        string city, string province);
+
+    public void TestLogging()
+    {
+        LogMethodSupportsPascalCasingOfNames("Vancouver", "BC");
+    }
+}
+```
+
+Consider the example logging output when using the `JsonConsole` formatter.
+
+```json
+{
+  "EventId": 13,
+  "LogLevel": "Information",
+  "Category": "LoggingExample",
+  "Message": "Welcome to Vancouver BC!",
+  "State": {
+    "Message": "Welcome to Vancouver BC!",
+    "City": "Vancouver",
+    "Province": "BC",
+    "{OriginalFormat}": "Welcome to {City} {Province}!"
+  }
+}
+```
+
+### Indeterminate parameter order
+
+There are no constraints on the ordering of log method parameters. A developer could define the `ILogger` as the last parameter, although it may a appear a bit awkward.
+
+```csharp
+[LoggerMessage(
+    EventId = 110,
+    Level = LogLevel.Debug,
+    Message = "M1 {ex3} {ex2}")]
+static partial void LogMethod(
+    Exception ex,
+    Exception ex2,
+    Exception ex3,
+    ILogger logger); // 😲
```

Comment:
I've never seen an emoji used in docs before. Style guide seems to say we shouldn't use them? https://styleguides.azurewebsites.net/Styleguide/Read?id=2700&topicid=33677

Reply:
```suggestion
    ILogger logger);
```